base/notify.go

```
@@ -0,0 +1,247 @@
+	Report Report `json:"report" yaml:"report"`
+
+	// JSONStringReport is the report but marshalled into JSON and stringified
+	JSONStringReport string `json:"JSONStringReport" yaml:"JSONStringReport"`
+
+	// EscapedJSONStringReport is the report but marshalled into JSON and stringified and with tabs and new lines escaped
+	EscapedJSONStringReport string `json:"escapedJSONStringReport" yaml:"escapedJSONStringReport"`
+}
+
+// getPayloadTemplateValues gets the values for the payload tempalte
+func getPayloadTemplateValues(exp *Experiment) (*NotifyPayloadTemplateValues, error) {
+	report := Report{
+		// Group: exp.driver.Group
+		// SLOs:           slos,
+		// Metrics:        metrics,
+
+		TimeStamp:         time.Now().String(),
+		Completed:         exp.Completed(),
+		NoTaskFailures:    exp.NoFailure(),
+		NumTasks:          len(exp.Spec),
+		NumCompletedTasks: exp.Result.NumCompletedTasks,
+		NumLoops:          exp.Result.NumLoops,
+		Experiment:        exp,
+	}
+
+	marshalledReport, err := json.MarshalIndent(report, "", "  ")
+	if err != nil {
+		log.Logger.Error("could not JSON marshall report")
+		return nil, err
+	}
+
+	stringReport := string(marshalledReport)
+
+	// escape double quotes, tabs, and new lines
+	escapedReport := strings.Replace(stringReport, "\"", "\\\"", -1)
+	escapedReport = strings.Replace(escapedReport, "\n", "\\n", -1)
+	escapedReport = strings.Replace(escapedReport, "\t", "\\t", -1)
+
+	return &NotifyPayloadTemplateValues{
+		Report:                  report,
+		JSONStringReport:        stringReport,
+		EscapedJSONStringReport: escapedReport,
+	}, nil
+}
+
+// getPayload fetches the payload template from the PayloadTemplateURL and
+// executes it with values from getPayloadTemplateValues()
+func (t *notifyTask) getPayload(exp *Experiment) (string, error) {
+	if t.With.PayloadTemplateURL != "" {
+		template, err := getProviderTemplate(t.With.PayloadTemplateURL)
+		if err != nil {
+			log.Logger.Error("could not get payload template")
```

This log here is not needed. The original `getProviderTemplate` should already do the logging.

---

Do you have any advice on how to determine if we should log something and when we should just return the error?

---

The first place where an error is discovered/created in Iter8 -- that's where logging needs to happen. For example, `getProvidereTemplate` already should log an error message (since it `knows` that it is returning an error), but if you Log that again in `notify.go`, the same log will appear twice.